### PR TITLE
Kafka-14748: Relax non-null FK left-join requirement

### PR DESCRIPTION
**Pull request by florin-akermann**
Relax non-null FK left-join requirement.

Testing Strategy: Inject extractor which returns null on first or second element.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by wcarlson5** [comment]
I'm going to start my review of this today. Hopefully can be actionable sometime tomorrow.

File: docs/streams/upgrade-guide.html [hunk 1]
```diff
@@ -198,6 +184,21 @@ <h3><a id="streams_api_changes_360" href="#streams_api_changes_360">Streams API
     </code>
     </pre>
     </p>
+
```
Comment (wcarlson5) [inline docs/streams/upgrade-guide.html:187]: Thanks for remembering the docs update!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java [hunk 2]
```diff
@@ -115,102 +116,93 @@ public void process(final Record<K, Change<V>> record) {
...
             }
         }
 
+        private byte[] serialize(final KO key) {
+            return foreignKeySerializer.serialize(foreignKeySerdeTopic, key);
+        }
+
+        private void forward(final Record<K, Change<V>> record, final KO foreignKey, final Instruction deleteKeyNoPropagate) {
+            final SubscriptionWrapper<K> wrapper = new SubscriptionWrapper<>(
+                hash(record),
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java:184]: How many times is this record rehased?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java [hunk 3]
```diff
@@ -115,102 +116,93 @@ public void process(final Record<K, Change<V>> record) {
...
-
-                final SubscriptionWrapper.Instruction instruction;
-                if (leftJoin) {
-                    //Want to send info even if RHS is null.
-                    instruction = PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE;
-                } else {
-                    instruction = PROPAGATE_ONLY_IF_FK_VAL_AVAILABLE;
-                }
-                final KO newForeignKey = foreignKeyExtractor.apply(record.value().newValue);
+                final KO newForeignKey = record.value().newValue == null ? null : foreignKeyExtractor.apply(record.value().newValue);
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java]: this seems redundant with line 165. That check the `newValue` is not `null`

**Review by wcarlson5 — COMMENTED** [review]
Just a couple of comments

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java (hunk 3 above)
Comment (florin-akermann) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java]: indeed!, thanks, removed.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java (hunk 2 above)
Comment (florin-akermann) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java:184]: there are branches where it is hashed twice. I puhsed a change to cache the hash value.

**Comment by florin-akermann** [comment]
> Just a couple of comments

Thank you @wcarlson5 for the comments.
I pushed according changes.

While I have your attention could we revisit the adjutments done to the optimization 'drop null key records on repartitioning' as part of https://github.com/apache/kafka/pull/14174#discussion_r1351880231.
To me it seems odd to apply the optimization in some cases and not in others.
E.g. as described in the linked comment, there could be this confusing case where null-key records don't get propagated because of a explicit repartition call at the end of the topology.

I was wondering whether it would make sense to adjust the optimization from 'filter null-key records in repartition nodes if no left or outer-join is downstream' to 'filter null-key records in repartition nodes if no left or outer-join node is part of this branch of the topology (aka no left or outer-join is up- or downstream of this repartition node)'

**Review by wcarlson5 — APPROVED** [review]

**Comment by wcarlson5** [comment]
I'm not sure, either option for the optimization is fine with me as long as it's well documented. 

I'm good with how the PR is for now. I'm going to merge it to get it in before feature freeze for 3.7

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java [hunk 4]
```diff
@@ -109,108 +111,102 @@ public void init(final ProcessorContext<KO, SubscriptionWrapper<K>> context) {
...
-
-            final int partition = context().recordMetadata().get().partition();
+        private void leftJoinInstructions(final Record<K, Change<V>> record) {
             if (record.value().oldValue != null) {
                 final KO oldForeignKey = foreignKeyExtractor.apply(record.value().oldValue);
+                final KO newForeignKey = record.value().newValue == null ? null : foreignKeyExtractor.apply(record.value().newValue);
+                if (oldForeignKey != null && !Arrays.equals(serialize(newForeignKey), serialize(oldForeignKey))) {
+                    forward(record, oldForeignKey, DELETE_KEY_AND_PROPAGATE);
+                }
+                forward(record, newForeignKey, PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java:136]: @florin-akermann @wcarlson5 -- Seems we introduces a bug here.

Filed: https://issues.apache.org/jira/browse/KAFKA-16644

Comment (florin-akermann) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplier.java:136]: @mjsax omg, thanks for the flag! Looks like @AyoubOm is addressing it in https://issues.apache.org/jira/browse/KAFKA-16394 already? Else i'll adress it.

**Pull request merged**